protocol/msgpack/unaggregated_iterator.go

```diff
@@ -0,0 +1,373 @@
+	if !ok {
+		return false
+	}
+	objType := it.decodeObjectType()
+	if it.err != nil {
+		return false
+	}
+	switch objType {
+	case counterWithPoliciesType, batchTimerWithPoliciesType, gaugeWithPoliciesType:
+		it.decodeMetricWithPolicies(objType)
+	default:
+		it.err = fmt.Errorf("unrecognized object type %v", objType)
+	}
+	it.skip(numActualFields - numExpectedFields)
+
+	return it.err == nil
+}
+
+func (it *unaggregatedIterator) decodeMetricWithPolicies(objType objectType) {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(objType)
+	if !ok {
+		return
+	}
+	switch objType {
+	case counterWithPoliciesType:
+		it.decodeCounter()
+	case batchTimerWithPoliciesType:
+		it.decodeBatchTimer()
+	case gaugeWithPoliciesType:
+		it.decodeGauge()
+	default:
+		it.err = fmt.Errorf("unrecognized metric with policies type %v", objType)
+		return
+	}
+	it.decodeVersionedPolicies()
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodeCounter() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(counterType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.CounterType
+	it.metric.ID = it.decodeID()
+	it.metric.CounterVal = int64(it.decodeVarint())
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodeBatchTimer() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(batchTimerType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.BatchTimerType
+	it.metric.ID = it.decodeID()
+	numValues := it.decodeArrayLen()
+	values := it.floatsPool.Get(numValues)
+	for i := 0; i < numValues; i++ {
+		values = append(values, it.decodeFloat64())
+	}
+	it.metric.BatchTimerVal = values
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodeGauge() {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(gaugeType)
+	if !ok {
+		return
+	}
+	it.metric.Type = unaggregated.GaugeType
+	it.metric.ID = it.decodeID()
+	it.metric.GaugeVal = it.decodeFloat64()
+	it.skip(numActualFields - numExpectedFields)
+}
+
+func (it *unaggregatedIterator) decodePolicy() policy.Policy {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(policyType)
+	if !ok {
+		return policy.Policy{}
+	}
+	resolution := it.decodeResolution()
+	retention := it.decodeRetention()
+	p := policy.Policy{Resolution: resolution, Retention: retention}
+	it.skip(numActualFields - numExpectedFields)
+	return p
+}
+
+func (it *unaggregatedIterator) decodeResolution() policy.Resolution {
+	numActualFields := it.decodeNumObjectFields()
+	resolutionType := it.decodeObjectType()
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
+		resolutionType,
+		numActualFields,
+	)
+	if !ok {
+		return policy.EmptyResolution
+	}
+	switch resolutionType {
+	case knownResolutionType:
+		resolutionValue := policy.ResolutionValue(it.decodeVarint())
```

Comment:
Is it worth making sure it is a known good enum value?  i.e. what if the int is out of range?